packages/babel-generator/src/generators/types.ts

```diff
@@ -188,44 +194,64 @@ export function NullLiteral(this: Printer) {
 }
 
 export function NumericLiteral(this: Printer, node: t.NumericLiteral) {
-  const raw = this.getPossibleRaw(node);
   const opts = this.format.jsescOption;
-  const value = node.value + "";
   if (opts.numbers) {
     this.number(jsesc(node.value, opts));
-  } else if (raw == null) {
-    this.number(value); // normalize
+    return;
+  }
+
+  const raw = this.getPossibleRaw(node);
+  if (raw == null) {
+    this.number(node.value + ""); // normalize
   } else if (this.format.minified) {
+    const value = node.value + "";
     this.number(raw.length < value.length ? raw : value);
   } else {
     this.number(raw);
   }
 }
 
-export function StringLiteral(this: Printer, node: t.StringLiteral) {
-  const raw = this.getPossibleRaw(node);
-  if (!this.format.minified && raw !== undefined) {
+export function StringLiteral(
+  this: Printer,
+  node: t.StringLiteral,
+  parent: NodePath<
+    t.FunctionDeclaration | t.FunctionExpression | t.TSDeclareFunction
+  >["parent"],
+) {
+  const inJsx = isJSXAttribute(parent);
+
+  const raw =
+    !this.format.minified &&
+    inJsx == !!node.extra?.jsx &&
+    this.getPossibleRaw(node);
+
+  if (typeof raw === "string") {
     this.token(raw);
     return;
   }
 
-  const val = jsesc(node.value, this.format.jsescOption);
+  const val = inJsx
+    ? `"${htmlEncode(node.value, {
+        level: EntityLevel.HTML,
+        mode: EncodingMode.UTF8,
+      })}"`
+    : jsesc(node.value, this.format.jsescOption);
```

Comment:
Is it safe to do just this, avoiding the extra dependency?

```suggestion
  let { value } = node.value;
  if (inJsx) value = value.replace(/&/g, "&amp;");
  value = jsesc(value, this.format.jsescOption);
```

Reply:
I'm not sure if `&` is necessary, `"` definitely needs this.
Also for `\`, it seems that jsx should not escape.